lib/statefileio.c

```diff
@@ -106,11 +106,41 @@ sf_FIST_delay_on_write();
 //
 
 
+enum {
+    DEVICE_TYPE_NORMAL = 0,
+    DEVICE_TYPE_NBD = 1
+};
+
+static int
+get_device_type(
+      const char *path)
+{
+    static const int nbd_major = 43;
+
+    struct stat stats;
+    if (stat(path, &stats) == -1)
+    {
+        log_internal(MTC_LOG_ERR, "SF: failed to execute stat (sys %d).\n", errno);
+        return DEVICE_TYPE_NORMAL;
+    }
+    if (S_ISBLK(stats.st_mode) && major(stats.st_rdev) == nbd_major) {
+        return DEVICE_TYPE_NBD;
+    }
+    return DEVICE_TYPE_NORMAL;
+}
+
 extern int
 sf_open(
     char *path)
 {
-    return open(path, (O_RDWR | O_DIRECT));
+    const int device_type = get_device_type(path);
+    const int flags = O_RDWR | O_DIRECT | (device_type == DEVICE_TYPE_NBD ? O_SYNC : 0);
```

Comment:
O_DIRECT and O_SYNC is a strange combination, O_SYNC would already wait for all data written to be written to the device, and we don't write that much data that buffering in memory would be a problem, so using O_SYNC alone should work (O_DIRECT also has additional requirements around alignment, and when they are not followed some filesystem drivers may simply fall back to buffered I/O anyway or return an error).